### PR TITLE
[18ZOO] Add A spoonful of sugar, A squeeze, Corn, On diet, Sparkling Gold, Tw…

### PR DESCRIPTION
**Pull request by carlorusso**
…o Barrels

### CODE REVIEW

File: lib/engine/game/base.rb [hunk 1]
```diff
@@ -1095,10 +1095,10 @@ def get(type, id)
         send("#{type}_by_id", id)
       end
 
-      def all_companies_with_ability(ability)
+      def all_companies_with_ability(ability_type)
```
Comment (carlorusso) [inline lib/engine/game/base.rb:1098]: As suggested by Michael B here: https://18xxgames.slack.com/archives/C012K0CNY5C/p1616541951268200?thread_ts=1616540448.268100&cid=C012K0CNY5C

**Comment by carlorusso** [comment]
@tobymao or anyone can check this?

File: lib/engine/game/g_18_zoo/ability/increase_distance_for_train.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Engine
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo/ability/increase_distance_for_train.rb:3]: why is this in 18zoo abilities specificially?

File: lib/engine/game/g_18_zoo/game.rb [hunk 3]
```diff
@@ -1310,6 +1336,12 @@ def a_spoonful_of_sugar
           @a_spoonful_of_sugar ||= company_by_id('A_SPOONFUL_OF_SUGAR')
         end
 
+        def can_choose_sugar?(entity, company)
+          company == a_spoonful_of_sugar && entity&.corporation? && a_spoonful_of_sugar.owner == entity &&
+            entity.trains.any? { |train| !%w[2J 4J].include? train.name } &&
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo/game.rb:1341]: always use parens

File: lib/engine/game/g_18_zoo/game.rb [hunk 4]
```diff
@@ -1337,13 +1369,15 @@ def log_share_price(entity, from, additional_info = '')
...
-          super
+          revenue = super
+
+          # Add 30$N if route contains 'Corn' and Corporation owns 'Corn'
+          revenue += 30 if route.corporation.assigned?(corn.id) && stops.any? { |stop| stop.hex.assigned?(corn.id) }
 
-          # TODO: Add 30$N if route contains 'Corn' and Corporation owns 'Corn'
-          # revenue += 30 if route.corporation.assigned?(corn.id) && stops.any? { |stop| stop.hex.assigned?(corn.id) }
+          # Towns revenues are doubled if 'Two barrels' is in use
+          revenue += 10 * stops.count { |stop| stop.tile.towns.any? } if two_barrels_used_this_or?(route.corporation)
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo/game.rb:1378]: !stop.title.towns.empty?

File: lib/engine/game/g_18_zoo/game.rb [hunk 5]
```diff
@@ -1361,22 +1395,90 @@ def check_distance(route, visits)
...
+
+          help << 'Routes get no subsidy at all, but every town increase route value by 10.' if barrel_assignment
+          help << 'Any train running in a city which has a Corn token increase route value by 30.' if corn_assignment
+          help
+        end
+
+        def subsidy_for(_route, stops)
+          subsidy = 0
+          # Get 1 for each town
+          subsidy += stops.count { |s| s.tile.towns.any? }
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo/game.rb:1440]: !empty?

File: lib/engine/game/g_18_zoo/step/assign.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+module Engine
+  module Game
+    module G18ZOO
+      module Step
+        class Assign < Engine::Step::Assign
+          def available_hex(entity, hex)
+            return unless entity.company?
+            return unless entity.owner&.corporation?
+            return unless entity.owner.tokens.map { |token| token&.city&.hex&.id }.include?(hex.id)
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo/step/assign.rb:11]: tokens.any? { |token| token&.city&.hex == hex )

File: lib/engine/game/g_18_zoo/step/choose_ability_on_or.rb [hunk 7]
```diff
@@ -151,12 +125,12 @@ def choices_for_zoo_ticket(company, corporation)
...
         # Not available for 2J / 4J, or if train has the 'Bandage'
-        # def choices_for_sugar(corporation)
-        #   corporation.trains
-        #              .reject { |train| %w[2J 4J].include? train.name }
-        #              .map { |train| [{ type: :sugar, train_id: train.id }, train.name] }
-        #              .to_h
-        # end
+        def choices_for_sugar(corporation)
+          corporation.trains
+                     .reject { |train| %w[2J 4J].include? train.name }
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo/step/choose_ability_on_or.rb:130]: parens

File: lib/engine/game/g_18_zoo/ability/increase_distance_for_train.rb (hunk 2 above)
Comment (carlorusso) [reply lib/engine/game/g_18_zoo/ability/increase_distance_for_train.rb:3]: I need to associate a train to the ability, so I added an ability here; should I use just “description” and keep the data inside game?

File: lib/engine/game/g_18_zoo/game.rb (hunk 3 above)
Comment (carlorusso) [reply lib/engine/game/g_18_zoo/game.rb]: done in d493f74b4

File: lib/engine/game/g_18_zoo/game.rb (hunk 4 above)
Comment (carlorusso) [reply lib/engine/game/g_18_zoo/game.rb]: done in d493f74b4

File: lib/engine/game/g_18_zoo/game.rb (hunk 5 above)
Comment (carlorusso) [reply lib/engine/game/g_18_zoo/game.rb]: done in d493f74b4

File: lib/engine/game/g_18_zoo/step/assign.rb (hunk 6 above)
Comment (carlorusso) [reply lib/engine/game/g_18_zoo/step/assign.rb]: done in d493f74b4

File: lib/engine/game/g_18_zoo/step/choose_ability_on_or.rb (hunk 7 above)
Comment (carlorusso) [reply lib/engine/game/g_18_zoo/step/choose_ability_on_or.rb]: done in d493f74b4

**Pull request merged**